Bottom: f7431f2dfbab602d5f34ebf4a990a1516c4c3129
Top:    29d4817221a31042fee32ab92c02174f34865665
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-04-17 07:46:51 +0800

tuntap: tweak on the patch of non-xdp case in tun_build_skb()

If we're sure not to go native XDP, there's no need for several things
like touching preemption counter and rcu stuffs.

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 2617e6f..9bbb8b5 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1641,10 +1641,12 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 	 * of xdp_prog above, this should be rare and for simplicity
 	 * we do XDP on skb in case the headroom is not enough.
 	 */
-	if (hdr->gso_type || !xdp_prog)
+	if (hdr->gso_type || !xdp_prog) {
 		*skb_xdp = 1;
-	else
-		*skb_xdp = 0;
+		goto build;
+	}
+
+	*skb_xdp = 0;
 
 	preempt_disable();
 	rcu_read_lock();
@@ -1690,6 +1692,7 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 	rcu_read_unlock();
 	preempt_enable();
 
+build:
 	skb = build_skb(buf, buflen);
 	if (!skb) {
 		skb = ERR_PTR(-ENOMEM);
